stestr/repository/sql.py

```diff
@@ -0,0 +1,246 @@
+        """Create a subunit2sql-based repository object for the repo at 'url'.
+
+        :param base: The path to the repository.
+        """
+        self.base = url
+        self.engine = sqlalchemy.create_engine(url)
+        self.session_factory = orm.sessionmaker(bind=self.engine,
+                                                autocommit=True)
+
+    # TODO(mtreinish): We need to add a subunit2sql api to get the count
+    def count(self):
+        super(Repository, self).count()
+
+    def _get_latest_run(self):
+        session = self.session_factory()
+        latest_run = db_api.get_latest_run(session)
+        session.close()
+        if not latest_run:
+            raise KeyError("No tests in repository")
+        return latest_run
+
+    def latest_id(self):
+        return self._get_latest_run().uuid
+
+    def get_failing(self):
+        latest_run = self._get_latest_run()
+        session = self.session_factory()
+        failed_test_runs = db_api.get_test_runs_by_status_for_run_ids(
+            'fail', [latest_run.id], session=session)
+        session.close()
+        return _Subunit2SqlRun(self.base, None, test_runs=failed_test_runs)
+
+    def get_test_run(self, run_id):
+        return _Subunit2SqlRun(self.base, run_id)
+
+    def _get_inserter(self, partial):
+        return _SqlInserter(self, partial)
+
+    def _get_test_times(self, test_ids):
+        result = {}
+        # TODO(mtreinish): after subunit2sql adds a bulk query for getting
+        # multiple tests by test_id at once remove the for loop
+        session = self.session_factory()
+        for test_id in test_ids:
+            test = db_api.get_test_by_test_id(test_id, session=session)
+            if test:
+                result[test_id] = test.run_time
+        session.close()
+        return result
+
+
+class _Subunit2SqlRun(repository.AbstractTestRun):
+    """A test run that was inserted into the repository."""
+
+    def __init__(self, url, run_id, test_runs=None):
+        engine = sqlalchemy.create_engine(url)
+        self.session_factory = orm.sessionmaker(bind=engine)
+        self._run_id = run_id
+        self._test_runs = test_runs
+
+    def get_id(self):
+        return self._run_id
+
+    def get_subunit_stream(self):
+        stream = io.BytesIO()
+        if self._run_id:
+            session = self.session_factory()
+            test_runs = db_api.get_tests_run_dicts_from_run_id(self._run_id,
+                                                               session)
+            session.close()
+        else:
+            test_runs = self._test_runs
+        output = subunit.v2.StreamResultToBytes(stream)
+        output.startTestRun()
+        for test_id in test_runs:
+            test = test_runs[test_id]
+            # NOTE(mtreinish): test_run_metadata is not guaranteed to be
+            # present for the test run
+            metadata = test.get('metadata', None)
+            write_subunit.write_test(output, test['start_time'],
```

Comment:
I suspect this is where the subunit stream is leaking to stdout, or something related to this call. From my quick scan, which is difficult to do from my phone (but so much easier than on gerrit) this gets called from get_test via inside run.py

But I'm not really sure, it's just a shot in the dark really. I'll take a more detailed look tomorrow when I'm in front of my computer again.